crates/vm/src/system/memory/online.rs

```diff
@@ -524,128 +600,203 @@ impl TracingMemory {
+                let (_, block_metadata) =
+                    self.get_block_metadata::<ALIGN>(address_space, split_ptr);
+                let timestamp = block_metadata.timestamp();
+                self.split_by_meta::<T, ALIGN>(
+                    split_ptr as u32,
+                    timestamp,
+                    split_size as u8,
+                    address_space,
+                );
             }
 
-            let prev_ts = (0..num_segs)
-                .map(|i| {
-                    let meta = self.meta[address_space].get(begin + i);
-                    if meta.block_size > 0 {
-                        meta.timestamp
-                    } else {
-                        // Initialize
-                        if self.initial_block_size >= ALIGN {
-                            // We need to split the initial block into chunks
-                            let block_start = (begin + i) & !(self.initial_block_size / ALIGN - 1);
-                            self.split_by_meta::<T, ALIGN>(
-                                &AccessMetadata {
-                                    start_ptr: (block_start * ALIGN) as u32,
-                                    block_size: self.initial_block_size as u32,
-                                    timestamp: INITIAL_TIMESTAMP,
-                                },
-                                address_space,
-                            );
-                        } else {
-                            debug_assert_eq!(self.initial_block_size, 1);
-                            debug_assert!((address_space as u32) < NATIVE_AS); // TODO: normal way
-                            self.add_merge_record(
-                                AccessRecordHeader {
-                                    timestamp_and_mask: INITIAL_TIMESTAMP,
-                                    address_space: address_space as u32,
-                                    pointer: (pointer + i * ALIGN) as u32,
-                                    block_size: ALIGN as u32,
-                                    lowest_block_size: self.initial_block_size as u32,
-                                    type_size: 1,
-                                },
-                                &INITIAL_CELL_BUFFER[..ALIGN], // TODO: this assumes cell_size=1
-                                &INITIAL_TIMESTAMP_BUFFER[..ALIGN],
-                            );
-                        }
-                        INITIAL_TIMESTAMP
-                    }
-                })
-                .collect::<Vec<_>>(); // PERF(AG): small buffer or small vec or something
+            // Process merge
+            let mut prev_ts_buf = [0u32; MAX_SEGMENTS];
+
+            let mut max_timestamp = INITIAL_TIMESTAMP;
+
+            let mut ptr = merge_ptr;
+            let end_ptr = merge_ptr + merge_size;
+            let mut seg_idx = 0;
+            while ptr < end_ptr {
+                let (_, block_metadata) = self.get_block_metadata::<ALIGN>(address_space, ptr);
```

Comment:
does this work properly now that `PagedVec::get` does not create the page? is it because you do `unwrap_or_default()` and then the page gets created in a later `set`?

Reply:
yes, i don't think we need to allocate pages if we're just fetching the values. i just return the default (zero) value. we only allocate pages on set